File: auth-server/src/main/java/no/vicx/authserver/TokenStatusController.java
package no.vicx.authserver;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService;
import org.springframework.security.oauth2.server.authorization.OAuth2TokenType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class TokenStatusController {

    private static final Logger LOG = LoggerFactory.getLogger(TokenStatusController.class);

    private final OAuth2AuthorizationService authorizationService;

    public TokenStatusController(OAuth2AuthorizationService authorizationService) {
        this.authorizationService = authorizationService;
    }

    @GetMapping("token-status")
    public ResponseEntity<?> tokenStatus(@RequestParam String token) {

        var authorization =
                authorizationService.findByToken(token, OAuth2TokenType.ACCESS_TOKEN);

        if (authorization != null) {
            var accessToken = authorization.getAccessToken();

            if (accessToken.isInvalidated()) {
                LOG.info("Token isInvalidated: {}", token);
                return ResponseEntity.ok("Token isInvalidated");
            } else {
                LOG.info("Token not invalidated: {}", token);
                return ResponseEntity.ok("Token not invalidated");
            }
        }

        LOG.warn("Non-existent token: {}", token);
        return ResponseEntity.notFound().build();
    }
}

File: auth-server/src/main/kotlin/no/vicx/authserver/TokenStatusController.kt
package no.vicx.authserver

import org.springframework.http.ResponseEntity
import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService
import org.springframework.security.oauth2.server.authorization.OAuth2TokenType
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
class TokenStatusController(
    private val authorizationService: OAuth2AuthorizationService,
) {
    @GetMapping("token-status")
    fun tokenStatus(
        @RequestParam token: String,
    ): ResponseEntity<*> {
        val authorization =
            authorizationService.findByToken(token, OAuth2TokenType.ACCESS_TOKEN)

        if (authorization == null) {
            log.warn("Non-existent token: {}", token)
            return ResponseEntity.notFound().build<Any>()
        }

        val accessToken = authorization.accessToken

        return if (accessToken.isInvalidated) {
            log.info("Token isInvalidated: {}", token)
            ResponseEntity.ok("Token isInvalidated")
        } else {
            log.info("Token not invalidated: {}", token)
            ResponseEntity.ok("Token not invalidated")
        }
    }

    companion object {
        private val log = loggerFor<TokenStatusController>()
    }
}
